File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        for i,e in enumerate(nums):

            if (i>0 and e == nums[i-1]):
                continue
            
            l = i + 1
            r = len(nums) - 1
            while (l < r):
                sumx = e + nums[l] + nums[r]
                if (sumx > 0):
                    r-= 1
                elif sumx < 0:
                    l+=1
                else:
                    res.append([e,nums[l],nums[r]])
                    l+=1
                    r-=1
                    while (nums[l] == nums[l-1] and l < r):
                        l+=1
        return res

        

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let newArr = new Array()
    nums.sort((a,b) => a-b)

    for (const [i,e] of nums.entries())
    {
        if (e > 0)
            break;
        
        if (i > 0 && e == nums[i-1])
            continue;

        let s = i+1
        let l = nums.length - 1
        while ( s < l )
        {
            let sumZero = nums[s] + nums[l] + e
            if (sumZero > 0)
                l -= 1
            else if (sumZero < 0)
                s += 1
            else{
                newArr.push([nums[s], nums[l], e])
                s += 1
                l -= 1
                while (nums[s] == nums[s -1] && s < l)
                    s+=1
            }
        }
    }
    return newArr
}